yamux/src/connection.rs

```diff
@@ -490,6 +501,32 @@ impl<T: AsyncRead + AsyncWrite + Unpin> Active<T> {
         }
     }
 
+    fn poll_stream_receivers(&mut self, cx: &mut Context) -> Poll<StreamCommand> {
+        for i in (0..self.stream_receivers.len()).rev() {
+            let (id, mut receiver) = self.stream_receivers.swap_remove(i);
```

Comment:
Polling each stream `Receiver` even though only one might be `Poll::Ready` seems wasteful. Would `SelectAll` not be an option? One would wrap each `Receiver` such that it returns the `StreamId` when the `Receiver` returns `Poll::Ready(None)`. With that `StreamId` we can then call `self.on_drop_stream(id)`.

(On consecutive polls the wrapper can return `Poll::Ready(None)` and thus it would be cleaned up by the `SelectAll`.)

Reply:
I had a `SelectAll` based design before but dropped it because I couldn't detect `None` from the outside.

Detecting that requires another data structure for translating between the stream commands and the actual frames in connection (the wrapping you mentioned).

This also needs to work for the `Cleanup` and `Close` case.

I tried hiding it in a custom collection object but we need access to the stream `IntMap` upon drop so that needs even more refactoring but would be a clean design.

I can invest the time if you want but I don't think it is a quick 30min refactoring.

Reply:
I was wrong, it was actually a super quick refactoring, now that I've seen how tokio's `StreamMap` works. They recommend an additional wrapper around `Stream`. The trick is to wrap the output of the `Stream` again such that you get a tuple of `(key, Option<Item>)` which allows you to detect closing of the stream.